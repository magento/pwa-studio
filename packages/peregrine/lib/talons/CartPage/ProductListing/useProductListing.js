import { useEffect, useState } from 'react';
import { useLazyQuery, useQuery } from '@apollo/client';

import { useCartContext } from '../../../context/cart';
import mergeOperations from '../../../util/shallowMerge';
import defaultOperations from './productListing.gql';

/**
 * This talon contains logic for a component that renders a list of products for a cart.
 * It performs effects and returns prop data to render the component on a cart page.
 *
 * This talon performs the following effects:
 *
 * - Fetch product listing data associated with the cart
 * - Log any GraphQL errors to the console
 *
 * @function
 *
 * @param {Object} props
 * @param {ProductListingQueries} props.queries GraphQL queries for getting product listing data.
 *
 * @returns {ProductListingTalonProps}
 *
 * @example <caption>Importing into your project</caption>
 * import { useProductListing } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useProductListing';
 */
export const useProductListing = props => {
    const {
        queries: { getProductListing }
    } = props;

    const operations = mergeOperations(defaultOperations, props.operations);

    const [{ cartId }] = useCartContext();
    const [activeEditItem, setActiveEditItem] = useState(null);

    const [
        fetchProductListing,
        { called, data, error, loading }
    ] = useLazyQuery(getProductListing, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first'
    });

    const { data: storeConfigData } = useQuery(
        operations.getWishlistConfigQuery
    );

    const wishlistConfig = storeConfigData ? storeConfigData.storeConfig : {};

    useEffect(() => {
        if (cartId) {
            fetchProductListing({
                variables: {
                    cartId
                }
            });
        }
    }, [cartId, fetchProductListing]);

    let items = [];
    if (called && !error && !loading) {
        items = data.cart.items;
    }

    return {
        activeEditItem,
        isLoading: !!loading,
        items,
        setActiveEditItem,
        wishlistConfig
    };
};

/** JSDocs type definitions */

/**
 * GraphQL queries for getting product listing data.
 * This is a type used in the {@link useProductListing} talon.
 *
 * @typedef {Object} ProductListingQueries
 *
 * @property {GraphQLDocument} getProductListing Query to get the product list for a cart
 *
 * @see [productListingFragments.js]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/ProductListing/productListingFragments.js}
 * for the queries used in Venia
 */

/**
 * Object type returned by the {@link useProductListing} talon.
 * It provides props data for a component that renders a product list.
 *
 * @typedef {Object} ProductListingTalonProps
 *
 * @property {Object} activeEditItem The product item currently being edited
 * @property {boolean} isLoading True if the query to get the product listing is still in progress. False otherwise.
 * @property {Array<Object>} items A list of products in a cart
 * @property {function} setActiveEditItem Function for setting the current item to edit
 *
 */
