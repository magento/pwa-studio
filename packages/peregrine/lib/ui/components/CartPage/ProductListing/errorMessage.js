import React from 'react';
import { shape, string } from 'prop-types';
import { useStyle } from '../../../classify';
import defaultClasses from './errorMessage.module.css';

const ErrorMessage = props => {
    const { error } = props;
    const classes = useStyle(defaultClasses, props.classes);

    if (!error || !error.message) {
        return null;
    }

    return <div className={classes.root}>{error.message}</div>;
};

ErrorMessage.defaultProps = {
    classes: null,
    error: null
};

ErrorMessage.propTypes = {
    classes: shape({
        root: string
    }),
    error: shape({
        message: string
    })
};

export default ErrorMessage;
