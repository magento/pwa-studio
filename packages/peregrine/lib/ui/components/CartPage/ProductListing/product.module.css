.root {
    composes: gap-y-2xs from global;
    composes: grid from global;
}

.item {
    composes: gap-x-2xs from global;
    composes: gap-y-xs from global;
    composes: grid from global;
    composes: items-start from global;
    grid-template-areas: 'image details kebab';
    grid-template-columns: 100px 1fr min-content;
}

.item_disabled {
    composes: item;
    composes: opacity-50 from global;
}

.errorText {
    composes: leading-normal from global;
    composes: text-error from global;
}

/* TODO @TW: cannot compose */
.errorText:empty {
    display: none;
}

.imageContainer {
    grid-area: image;
}

.imageRoot {
    composes: h-full from global;
}

.image {
    composes: bg-subtle from global;
    composes: border from global;
    composes: border-solid from global;
    composes: border-subtle from global;
    composes: h-full from global;
    composes: object-contain from global;
    composes: object-center from global;
    composes: rounded-sm from global;
}

.details {
    composes: gap-2xs from global;
    composes: leading-normal from global;

    composes: sm_grid from global;
}

@media (min-width: 640px) {
    .details {
        grid-area: details;
        grid-template-areas:
            'name name'
            'options quantity'
            'price quantity'
            'stock quantity';
        grid-template-columns: 2fr 1fr;
    }
}

.name {
    composes: font-semibold from global;
    grid-area: name;
}

.price {
    composes: text-sm from global;
    grid-area: price;
}

.quantity {
    composes: grid from global;
    composes: items-start from global;
    composes: justify-items-start from global;
    grid-area: quantity;

    composes: sm_justify-items-center from global;
}

.kebab {
    composes: relative from global;
    grid-area: kebab;
}

.sectionText {
    composes: pointer-events-none from global;
    composes: px-2xs from global;
    composes: py-1 from global;
    composes: text-sm from global;
}

.options {
    composes: gap-0.5 from global;
    composes: grid from global;
    composes: text-sm from global;
    grid-area: options;
}

.optionLabel {
    composes: auto-cols-max from global;
    composes: grid from global;
    composes: grid-flow-col from global;
}

.stockStatusMessage {
    composes: font-semibold from global;
    composes: text-error from global;
    composes: text-sm from global;
    grid-area: stock;
}

/* TODO @TW: cannot compose */
.stockStatusMessage:empty {
    display: none;
}

.addToListButton {
    --stroke: rgb(var(--venia-global-color-teal));

    composes: content-center from global;
    composes: gap-x-xs from global;
    composes: inline-flex from global;
    composes: px-2.5 from global;
    composes: py-3.5 from global;
    composes: text-sm from global;
    composes: w-full from global;
}

.addToListButton_selected {
    composes: hidden from global;
}

@media (max-width: 639px) {
    .name,
    .options,
    .price,
    .stockStatusMessage,
    .quantity {
        grid-area: auto;
    }
}
