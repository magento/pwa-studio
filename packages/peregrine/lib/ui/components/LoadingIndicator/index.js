export { default } from './indicator';
export { default as fullPageLoadingIndicator } from './static';
export { default as Spinner } from './spinner';
