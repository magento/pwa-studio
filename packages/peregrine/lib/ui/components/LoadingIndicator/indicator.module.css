.root {
    composes: content-center from global;
    composes: gap-lg from global;
    composes: grid from global;
    composes: items-center from global;
    composes: justify-center from global;
    composes: justify-items-center from global;
    height: calc(100% - 6rem);
    composes: mx-0 from global;
    composes: my-lg from global;
    composes: w-full from global;
}

.global {
    composes: root;

    composes: fixed from global;
    composes: h-full from global;
    composes: left-0 from global;
    composes: m-0 from global;
    composes: top-0 from global;
    composes: w-full from global;
}

.message {
    composes: text-colorDefault from global;
    composes: text-subtle from global;
    composes: uppercase from global;
    letter-spacing: 0.25em;
}

.indicator {
    animation-direction: alternate;
    animation-duration: 1.5s;
    animation-iteration-count: infinite;
    animation-name: pulse;
    animation-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
}

@keyframes pulse {
    0% {
        filter: grayscale(1);
        transform: scale(0.75) rotate(-540deg);
    }
    100% {
        filter: grayscale(0);
        transform: scale(1) rotate(0deg);
    }
}
