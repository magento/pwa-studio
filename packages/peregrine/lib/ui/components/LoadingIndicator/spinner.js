import React from 'react';

import defaultClasses from './spinner.module.css';
import { useStyle } from '../../classify';
import { RotateCw as LoaderIcon } from 'react-feather';
import Icon from '../Icon';

const Spinner = props => {
    const classes = useStyle(defaultClasses, props.classes);

    return (
        <Icon
            src={LoaderIcon}
            size={24}
            classes={{ root: classes.root, icon: classes.icon }}
        />
    );
};

export default Spinner;
