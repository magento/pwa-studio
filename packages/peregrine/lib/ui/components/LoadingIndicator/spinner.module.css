.root {
    composes: root from '../Icon/icon.module.css';
    --stroke: rgb(var(--venia-global-color-text-hint));
}

.icon {
    composes: icon from '../Icon/icon.module.css';
    composes: animation-spin from global;
}
