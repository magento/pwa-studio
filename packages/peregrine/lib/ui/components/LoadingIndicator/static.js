import React from 'react';
import { FormattedMessage } from 'react-intl';

import LoadingIndicator from './indicator';

const staticIndicator = (
    <LoadingIndicator global={true}>
        <FormattedMessage
            id={'loadingIndicator.message'}
            defaultMessage={'Fetching Data...'}
        />
    </LoadingIndicator>
);

export default staticIndicator;
