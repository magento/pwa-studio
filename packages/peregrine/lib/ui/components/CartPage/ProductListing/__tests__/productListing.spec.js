import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import LoadingIndicator from '../../../LoadingIndicator';
import ProductListing from '../productListing';
import { useProductListing } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useProductListing';

jest.mock(
    '@magento/peregrine/lib/talons/CartPage/ProductListing/useProductListing'
);
jest.mock('../../../../classify');
jest.mock('@apollo/client', () => {
    return {
        gql: jest.fn(),
        useLazyQuery: jest.fn()
    };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});
jest.mock('react-router-dom', () => ({
    Link: ({ children, ...rest }) => <div {...rest}>{children}</div>
}));
jest.mock('@magento/peregrine/lib/util/makeUrl');
jest.mock('../product', () => 'Product');
jest.mock('../EditModal', () => 'EditModal');

test('renders null with no items in cart', () => {
    useProductListing.mockReturnValueOnce({
        isLoading: false,
        isUpdating: false,
        items: []
    });

    const tree = createTestInstance(<ProductListing />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders list of products with items in cart', () => {
    useProductListing.mockReturnValueOnce({
        isLoading: false,
        isUpdating: false,
        items: ['1', '2', '3']
    });

    const tree = createTestInstance(<ProductListing />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders loading indicator if isLoading', () => {
    useProductListing.mockReturnValueOnce({
        isLoading: true
    });

    const propsWithClass = {
        classes: {
            root: 'root'
        }
    };

    const tree = createTestInstance(<ProductListing {...propsWithClass} />);

    expect(tree.root.findByType(LoadingIndicator)).toBeTruthy();
});
