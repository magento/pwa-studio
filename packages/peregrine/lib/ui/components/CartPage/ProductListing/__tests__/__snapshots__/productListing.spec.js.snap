// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders list of products with items in cart 1`] = `
Array [
  <ul
    className="root"
  >
    <Product
      item="1"
    />
    <Product
      item="2"
    />
    <Product
      item="3"
    />
  </ul>,
  <require('./EditModal') />,
]
`;

exports[`renders null with no items in cart 1`] = `null`;
